File: packages/uns-py/src/uns_kit/core/service_token_provider.py
from __future__ import annotations

import os
from pathlib import Path
from typing import Callable, Mapping, Optional, Protocol


class AccessTokenProvider(Protocol):
    """Minimal bearer-token source shared by service clients."""

    def get_access_token(self) -> Optional[str]: ...


class ServiceTokenProvider:
    """Resolve controller-managed service credentials without storing a user token.

    A controller-mounted token file is deliberately authoritative and is read on
    every request, so a controller can rotate the service token atomically
    without restarting its RTT process. Direct local development can instead
    supply ``UNS_SERVICE_TOKEN``, a resolved ``uns.token``, or a legacy login
    provider.
    """

    def __init__(
        self,
        *,
        token_file: Optional[str] = None,
        environment: Optional[Mapping[str, str]] = None,
        config_token: Optional[str] = None,
        fallback: Optional[AccessTokenProvider] = None,
        read_text: Optional[Callable[[str], str]] = None,
    ) -> None:
        self._token_file = _normalize_token(token_file)
        self._environment = environment if environment is not None else os.environ
        self._config_token = _normalize_token(config_token)
        self._fallback = fallback
        self._read_text = read_text or _read_token_file

    def get_access_token(self) -> Optional[str]:
        token_file = self._token_file or _normalize_token(self._environment.get("UNS_SERVICE_TOKEN_FILE"))
        if token_file:
            return self._read_controller_managed_token(token_file)

        environment_token = _normalize_token(self._environment.get("UNS_SERVICE_TOKEN"))
        if environment_token:
            return environment_token

        if self._config_token:
            return self._config_token

        return self._fallback.get_access_token() if self._fallback else None

    def _read_controller_managed_token(self, path: str) -> str:
        try:
            token = _normalize_token(self._read_text(path))
            if not token:
                raise ValueError("empty token file")
            return token
        except Exception as exc:
            raise RuntimeError("Controller-managed service token file is unavailable.") from exc


def _read_token_file(path: str) -> str:
    return Path(path).read_text(encoding="utf-8")


def _normalize_token(value: Optional[str]) -> Optional[str]:
    token = value.strip() if isinstance(value, str) else ""
    return token or None

File: packages/uns-core/src/tools/auth/service-token-provider.ts
import { readFile as readFileFromFs } from "node:fs/promises";

/**
 * A source of bearer tokens for controller and DataHub requests.
 *
 * The interface is intentionally small so services can use a controller-issued
 * machine token without depending on the legacy user-login client.
 */
export interface AccessTokenProvider {
  getAccessToken(): Promise<string | undefined>;
}

type ServiceTokenEnvironment = Partial<Pick<NodeJS.ProcessEnv, "UNS_SERVICE_TOKEN_FILE" | "UNS_SERVICE_TOKEN">>;

export type ServiceTokenProviderOptions = {
  /**
   * Path to a controller-managed token file. When set, the file is
   * authoritative and failures are reported instead of falling back to another
   * credential source.
   */
  tokenFile?: string;
  /** Environment used to resolve UNS_SERVICE_TOKEN_FILE and UNS_SERVICE_TOKEN. */
  environment?: ServiceTokenEnvironment;
  /** Resolved config.uns.token value. */
  configToken?: string;
  /** Legacy development fallback, normally an AuthClient using uns.email/password. */
  fallback?: AccessTokenProvider;
  /** Test seam; production uses node:fs/promises.readFile. */
  readFile?: (path: string, encoding: BufferEncoding) => Promise<string>;
};

/**
 * Resolves service credentials in the order used by controller-managed RTT
 * services: mounted token file, environment, configuration, then the legacy
 * login provider. The mounted file is deliberately read for each request so
 * atomically-rotated controller credentials are picked up without restart.
 */
export class ServiceTokenProvider implements AccessTokenProvider {
  private readonly tokenFile?: string;
  private readonly environment: ServiceTokenEnvironment;
  private readonly configToken?: string;
  private readonly fallback?: AccessTokenProvider;
  private readonly readFile: (path: string, encoding: BufferEncoding) => Promise<string>;

  constructor(options: ServiceTokenProviderOptions = {}) {
    this.tokenFile = normalizeToken(options.tokenFile);
    this.environment = options.environment ?? process.env;
    this.configToken = normalizeToken(options.configToken);
    this.fallback = options.fallback;
    this.readFile = options.readFile ?? readFileFromFs;
  }

  async getAccessToken(): Promise<string | undefined> {
    const tokenFile = this.tokenFile ?? normalizeToken(this.environment.UNS_SERVICE_TOKEN_FILE);
    if (tokenFile) {
      return this.readControllerManagedToken(tokenFile);
    }

    const environmentToken = normalizeToken(this.environment.UNS_SERVICE_TOKEN);
    if (environmentToken) {
      return environmentToken;
    }

    if (this.configToken) {
      return this.configToken;
    }

    return this.fallback?.getAccessToken();
  }

  private async readControllerManagedToken(path: string): Promise<string> {
    try {
      const token = normalizeToken(await this.readFile(path, "utf8"));
      if (!token) {
        throw new Error("empty token file");
      }
      return token;
    } catch (error) {
      throw new Error("Controller-managed service token file is unavailable.", {
        cause: error instanceof Error ? error : undefined,
      });
    }
  }
}

function normalizeToken(value: string | undefined): string | undefined {
  const token = value?.trim();
  return token ? token : undefined;
}
